source:generated within constraint.yaml causes gator verify to fail

**What steps did you take and what happened:**
I'm using expansionTemplates for the gatekeeper-library policies i'm importing. I'm explicitly setting `spec.match.source: "Generated"` on the constraint.yaml file. I am also using gator verify for testing. I'm having issues where I set the source: "Generated" and my gator verify fails. When i remove source: "Generated" from the constraint.yaml it passes. 

**Failed test:**
```
> gator verify  opa/tests/...             
    --- FAIL: disallowed        (0.003s)
        unexpected number of violations: got 0 violations but want at least 1: got messages []
--- FAIL: forbidden-sysctls     (0.008s)
FAIL    opa/tests/forbidden-sysctls/suite.yaml  0.008s
```


**What did you expect to happen:**
I expected the `gator verify opa/tests/...` to pass. 

**constraint.yaml:**
```
apiVersion: constraints.gatekeeper.sh/v1beta1
kind: K8sPSPForbiddenSysctls
metadata:
  name: k8spspforbiddensysctls
spec:
  enforcementAction: warn
  match:
    excludedNamespaces:
      - gatekeeper
      - kube-system
    kinds:
      - apiGroups:
          - ''
        kinds:
          - Pod
    source: Generated
  parameters:
    allowedSysctls:
      - vm.max_map_count
    forbiddenSysctls: []
```
**template.yaml**
```
apiVersion: templates.gatekeeper.sh/v1
kind: ConstraintTemplate
metadata:
  name: k8spspforbiddensysctls
  annotations:
    metadata.gatekeeper.sh/title: Forbidden sysctls
    metadata.gatekeeper.sh/version: 1.0.0
    description: |
      Controls the `sysctl` profile used by containers. Corresponds to the
      `allowedUnsafeSysctls` and `forbiddenSysctls` fields in a PodSecurityPolicy.
      When specified, any sysctl not in the `allowedSysctls` parameter is considered to be forbidden.
      The `forbiddenSysctls` parameter takes precedence over the `allowedSysctls` parameter.
      For more information, see https://kubernetes.io/docs/tasks/administer-cluster/sysctl-cluster/
spec:
  crd:
    spec:
      names:
        kind: K8sPSPForbiddenSysctls
      validation:
        openAPIV3Schema:
          type: object
          properties:
            allowedSysctls:
              type: array
              description: An allow-list of sysctls. `*` allows all sysctls not listed in the `forbiddenSysctls` parameter.
              items:
                type: string
            forbiddenSysctls:
              type: array
              description: A disallow-list of sysctls. `*` forbids all sysctls.
              items:
                type: string
  targets:
    - target: admission.k8s.gatekeeper.sh
      rego: |
        package k8spspforbiddensysctls

        import data.lib.exclude_update.is_update

        # Block if forbidden
        violation[{"msg": msg, "details": {}}] {
        	# spec.securityContext.sysctls field is immutable.
        	not is_update(input.review)

        	sysctl := input.review.object.spec.securityContext.sysctls[_].name
        	forbidden_sysctl(sysctl)
        	msg := sprintf("The sysctl %v is not allowed, pod: %v. Forbidden sysctls: %v", [sysctl, input.review.object.metadata.name, input.parameters.forbiddenSysctls])
        }

        # Block if not explicitly allowed
        violation[{"msg": msg, "details": {}}] {
        	not is_update(input.review)
        	sysctl := input.review.object.spec.securityContext.sysctls[_].name
        	not allowed_sysctl(sysctl)
        	msg := sprintf("The sysctl %v is not explicitly allowed, pod: %v. Allowed sysctls: %v", [sysctl, input.review.object.metadata.name, input.parameters.allowedSysctls])
        }

        # * may be used to forbid all sysctls
        forbidden_sysctl(_) {
        	input.parameters.forbiddenSysctls[_] == "*"
        }

        forbidden_sysctl(sysctl) {
        	input.parameters.forbiddenSysctls[_] == sysctl
        }

        forbidden_sysctl(sysctl) {
        	forbidden := input.parameters.forbiddenSysctls[_]
        	endswith(forbidden, "*")
        	startswith(sysctl, trim_suffix(forbidden, "*"))
        }

        # * may be used to allow all sysctls
        allowed_sysctl(_) {
        	input.parameters.allowedSysctls[_] == "*"
        }

        allowed_sysctl(sysctl) {
        	input.parameters.allowedSysctls[_] == sysctl
        }

        allowed_sysctl(sysctl) {
        	allowed := input.parameters.allowedSysctls[_]
        	endswith(allowed, "*")
        	startswith(sysctl, trim_suffix(allowed, "*"))
        }
      libs:
        - |
          package lib.exclude_update

          is_update(review) {
          	review.operation == "UPDATE"
          }
```

**suite.yaml:**
```
kind: Suite
apiVersion: test.gatekeeper.sh/v1alpha1
metadata:
  name: forbidden-sysctls
tests:
- name: forbidden-sysctls
  template: ../../general/forbidden-sysctls/template.yaml
  constraint: ../../general/forbidden-sysctls/constraint.yaml
  cases:
  - name: allowed
    object: samples/allowed.yaml
    assertions:
    - violations: no
  - name: disallowed
    object: samples/disallowed.yaml
    assertions:
    - violations: yes
```

**disallowed.yaml:**
```
apiVersion: v1
kind: Pod
metadata:
  name: sample-container
spec:
  securityContext:
    sysctls:
      - name: vm.max_map_count
        value: "242144"
      - name: test_sysctl
        value: "65536"
  containers:
    - name: sample-container
      image: busybox:latest
```

allowed:
```
apiVersion: v1
kind: Pod
metadata:
  name: sample-container
spec:
  securityContext:
    sysctls:
      - name: vm.max_map_count
        value: "242144"
  containers:
    - name: sample-container
      image: busybox:latest
```
expansionTemplate
```
apiVersion: expansion.gatekeeper.sh/v1alpha1
kind: ExpansionTemplate
metadata:
  name: expand-deployments
spec:
  applyTo:
    - groups: ["apps"]
      kinds: ["DaemonSet", "Deployment", "ReplicaSet", "StatefulSet"]
      versions: ["v1"]
    - groups: [""]
      kinds: ["ReplicationController"]
      versions: ["v1"]
    - groups: ["batch"]
      kinds: ["Job"]
      versions: ["v1"]
  templateSource: "spec.template"
  generatedGVK:
    kind: "Pod"
    group: ""
    version: "v1"
```

- Gatekeeper version: v3.15.1
- Kubernetes version: (use `kubectl version`):  v1.29.4

> > The source field on the match API, present in the Mutation and Constraint kinds, specifies if the config should match Generated ( i.e. fake) resources, Original resources, or both. The source field is an enum which accepts the following values:
> > Generated – the config will only apply to expanded resources, and will not apply to any real resources on the cluster
> 
> https://open-policy-agent.github.io/gatekeeper/website/docs/expansion
> 
> In your test suite, the pod yaml is not a fake resource.
> 
> When you remove `Generated` from the constraint resource, it worked because:
> 
> > All – the config will apply to both Generated and Original resources. This is the default value.



> > The source field on the match API, present in the Mutation and Constraint kinds, specifies if the config should match Generated ( i.e. fake) resources, Original resources, or both. The source field is an enum which accepts the following values:
> > Generated – the config will only apply to expanded resources, and will not apply to any real resources on the cluster
> 
> https://open-policy-agent.github.io/gatekeeper/website/docs/expansion
> 
> In your test suite, the pod yaml is not a fake resource.
> 
> When you remove `Generated` from the constraint resource, it worked because:
> 
> > All – the config will apply to both Generated and Original resources. This is the default value.

ok i changed the allowed and disallowed.yaml to a deployment and its still failing:

```
> gator verify opa/tests/... 
    --- FAIL: disallowed        (0.003s)
        unexpected number of violations: got 0 violations but want at least 1: got messages []
--- FAIL: forbidden-sysctls     (0.009s)
FAIL    opa/tests/forbidden-sysctls/suite.yaml  0.009s

Error: FAIL
```

allowed.yaml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello
spec:
  replicas: 1
  selector:
    matchLabels:
      app: hello
  template:
    metadata:
      labels:
        app: hello
    spec:
      securityContext:
        sysctls:
        - name: vm.max_map_count
          value: "242144"
      containers:
        - name: hello
          image: busybox
          command: ["sh", "-c"]
          args:
            - sleep 36010

```
disallowed.yaml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello
spec:
  replicas: 1
  selector:
    matchLabels:
      app: hello
  template:
    metadata:
      labels:
        app: hello
    spec:
      securityContext:
        capabilities:
          add:
          - SYS_ADMIN
        sysctls:
        - name: test
          value: "1024"
      containers:
        - name: hello
          image: busybox
          command: ["sh", "-c"]
          args:
            - sleep 36010
 ```
 constraint.yaml
```
apiVersion: constraints.gatekeeper.sh/v1beta1
kind: K8sPSPForbiddenSysctls
metadata:
  name: k8spspforbiddensysctls
spec:
  enforcementAction: warn
  match:
    excludedNamespaces:
      - gatekeeper
      - kube-system
    kinds:
      - apiGroups:
          - ''
        kinds:
          - Pod
    source: Generated
  parameters:
    allowedSysctls:
      - vm.max_map_count
    forbiddenSysctls: []

```

--- a/pkg/gator/errors.go
+++ b/pkg/gator/errors.go
@@ -15,10 +15,9 @@ var (
 	// ErrNotASyncSet indicates the user-indicated file does not contain a
 	// SyncSet.
 	ErrNotASyncSet = errors.New("not a SyncSet")
-	// ErrNotASyncSet indicates the user-indicated file does not contain a
-	// SyncSet.
+	// ErrNotAGVKManifest indicates the user-indicated file does not contain a
+	// GVK Manifest.
 	ErrNotAGVKManifest = errors.New("not a GVKManifest")
-	// ErrNotAnExpansion indicates the user-indicated file does not contain a
 	// ErrNotAnExpansion indicates the user-indicated file does not contain an
 	// ExpansionTemplate.
 	ErrNotAnExpansion = errors.New("not an Expansion Template")

--- a/pkg/gator/errors.go
+++ b/pkg/gator/errors.go
@@ -18,6 +18,9 @@ var (
 	// ErrNotASyncSet indicates the user-indicated file does not contain a
 	// SyncSet.
 	ErrNotAGVKManifest = errors.New("not a GVKManifest")
+	// ErrNotAnExpansion indicates the user-indicated file does not contain a
+	// Constraint.
+	ErrNotAnExpansion = errors.New("not an Expansion Template")
 	// ErrAddingTemplate indicates a problem instantiating a Suite's ConstraintTemplate.
 	ErrAddingTemplate = errors.New("adding template")
 	// ErrAddingConstraint indicates a problem instantiating a Suite's Constraint.

--- a/pkg/gator/errors.go
+++ b/pkg/gator/errors.go
@@ -19,7 +19,7 @@ var (
 	// SyncSet.
 	ErrNotAGVKManifest = errors.New("not a GVKManifest")
 	// ErrNotAnExpansion indicates the user-indicated file does not contain a
-	// Constraint.
+	// ExpansionTemplate.
 	ErrNotAnExpansion = errors.New("not an Expansion Template")
 	// ErrAddingTemplate indicates a problem instantiating a Suite's ConstraintTemplate.
 	ErrAddingTemplate = errors.New("adding template")

--- a/pkg/gator/errors.go
+++ b/pkg/gator/errors.go
@@ -19,6 +19,7 @@ var (
 	// SyncSet.
 	ErrNotAGVKManifest = errors.New("not a GVKManifest")
 	// ErrNotAnExpansion indicates the user-indicated file does not contain a
+	// ErrNotAnExpansion indicates the user-indicated file does not contain an
 	// ExpansionTemplate.
 	ErrNotAnExpansion = errors.New("not an Expansion Template")
 	// ErrAddingTemplate indicates a problem instantiating a Suite's ConstraintTemplate.

--- a/pkg/gator/fixtures/fixtures.go
+++ b/pkg/gator/fixtures/fixtures.go
@@ -158,6 +158,34 @@ spec:
         }
 `
 
+	TemplateRestrictCustomField = `
+kind: ConstraintTemplate
+apiVersion: templates.gatekeeper.sh/v1beta1
+metadata:
+  name: restrictedcustomfield
+spec:
+  crd:
+    spec:
+      names:
+        kind: RestrictedCustomField
+      validation:
+        openAPIV3Schema:
+          type: object
+          properties:
+            expectedCustomField: 
+              type: boolean
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package restrictedcustomfield
+        violation[{"msg": msg}] {
+          got := input.review.object.spec.customField
+          expected := input.parameters.expectedCustomField
+          got == expected
+          msg := sprintf("foo object has restricted custom field value of %v", [expected])
+        }
+`
+
 	ConstraintAlwaysValidate = `
 kind: AlwaysValidate
 apiVersion: constraints.gatekeeper.sh/v1beta1
@@ -262,6 +290,22 @@ metadata:
   name: other
 `
 
+	ConstraintRestrictCustomField = `
+apiVersion: constraints.gatekeeper.sh/v1beta1
+kind: RestrictedCustomField
+metadata:
+  name: restrict-foo-custom-field
+spec:
+  match:
+    kinds:
+      - apiGroups: [""]
+        kinds: ["Foo"]
+    namespaces:
+      - "default"
+  parameters:
+    expectedCustomField: true
+`
+
 	Object = `
 kind: Object
 apiVersion: group.sh/v1
@@ -327,6 +371,17 @@ kind: Object
 apiVersion: group.sh/v1
 metadata:
   name: object`
+
+	ObjectFooTemplate = `
+apiVersion: apps/v1
+kind: FooTemplate
+metadata:
+  name: foo-template
+spec:
+  template:
+    spec:
+      customField: true
+`
 
 	NamespaceSelected = `
 kind: Namespace
@@ -681,5 +736,22 @@ spec:
     kinds:
       - apiGroups: ["*"]
         kinds: ["*"]
+`
+
+	ExpansionRestrictCustomField = `
+apiVersion: expansion.gatekeeper.sh/v1alpha1
+kind: ExpansionTemplate
+metadata:
+  name: expand-foo
+spec:
+  applyTo:
+  - groups: [ "apps" ]
+    kinds: [ "FooTemplate" ]
+    versions: [ "v1" ]
+  templateSource: "spec.template"
+  generatedGVK:
+    kind: "Foo"
+    group: ""
+    version: "v1"
 `
 )

--- a/pkg/gator/reader/read_resources.go
+++ b/pkg/gator/reader/read_resources.go
@@ -243,6 +243,20 @@ func ReadConstraint(f fs.FS, path string) (*unstructured.Unstructured, error) {
 	return u, nil
 }
 
+func ReadExpansion(f fs.FS, path string) (*unstructured.Unstructured, error) {
+	u, err := ReadObject(f, path)
+	if err != nil {
+		return nil, err
+	}
+
+	gvk := u.GroupVersionKind()
+	if gvk.Group != "expansion.gatekeeper.sh" {
+		return nil, gator.ErrNotAnExpansion
+	}
+
+	return u, nil
+}
+
 // ReadK8sResources reads JSON or YAML k8s resources from an io.Reader,
 // decoding them into Unstructured objects and returning those objects as a
 // slice.

--- a/pkg/gator/reader/read_resources.go
+++ b/pkg/gator/reader/read_resources.go
@@ -250,7 +250,7 @@ func ReadExpansion(f fs.FS, path string) (*unstructured.Unstructured, error) {
 	}
 
 	gvk := u.GroupVersionKind()
-	if gvk.Group != "expansion.gatekeeper.sh" {
+	if gvk.Group != "expansion.gatekeeper.sh" || gvk.Kind != "ExpansionTemplate" {
 		return nil, gator.ErrNotAnExpansion
 	}
 

--- a/pkg/gator/verify/runner.go
+++ b/pkg/gator/verify/runner.go
@@ -125,9 +125,7 @@ func (r *Runner) runTest(ctx context.Context, suiteDir string, filter Filter, t 
 	start := time.Now()
 
 	err := r.tryAddConstraint(ctx, suiteDir, t)
-	err = r.tryAddExpansion(suiteDir, t)
 	var results []CaseResult
-	// What is this Invalid and where does it get set? I didn't see it get set in tryAddConstraints
 	if t.Invalid {
 		if errors.Is(err, constraints.ErrSchema) {
 			err = nil
@@ -195,6 +193,11 @@ func (r *Runner) runCases(ctx context.Context, suiteDir string, filter Filter, t
 		}
 
 		return e, nil
+	}
+
+	_, err := newExpander()
+	if err != nil {
+		return nil, err
 	}
 
 	results := make([]CaseResult, len(t.Cases))

--- a/pkg/gator/verify/runner.go
+++ b/pkg/gator/verify/runner.go
@@ -12,7 +12,9 @@ import (
 	"github.com/open-policy-agent/frameworks/constraint/pkg/client/reviews"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/types"
 	"github.com/open-policy-agent/gatekeeper/v3/apis"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/expansion"
 	"github.com/open-policy-agent/gatekeeper/v3/pkg/gator"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/gator/expand"
 	"github.com/open-policy-agent/gatekeeper/v3/pkg/gator/reader"
 	mutationtypes "github.com/open-policy-agent/gatekeeper/v3/pkg/mutation/types"
 	"github.com/open-policy-agent/gatekeeper/v3/pkg/target"
@@ -123,7 +125,9 @@ func (r *Runner) runTest(ctx context.Context, suiteDir string, filter Filter, t 
 	start := time.Now()
 
 	err := r.tryAddConstraint(ctx, suiteDir, t)
+	err = r.tryAddExpansion(suiteDir, t)
 	var results []CaseResult
+	// What is this Invalid and where does it get set? I didn't see it get set in tryAddConstraints
 	if t.Invalid {
 		if errors.Is(err, constraints.ErrSchema) {
 			err = nil
@@ -140,6 +144,11 @@ func (r *Runner) runTest(ctx context.Context, suiteDir string, filter Filter, t 
 		Runtime:     Duration(time.Since(start)),
 		CaseResults: results,
 	}
+}
+
+func (r *Runner) tryAddExpansion(suiteDir string, t *Test) error {
+	_, err := r.makeTestExpander(suiteDir, t)
+	return err
 }
 
 func (r *Runner) tryAddConstraint(ctx context.Context, suiteDir string, t *Test) error {
@@ -179,6 +188,15 @@ func (r *Runner) runCases(ctx context.Context, suiteDir string, filter Filter, t
 		return c, nil
 	}
 
+	newExpander := func() (*expand.Expander, error) {
+		e, err := r.makeTestExpander(suiteDir, t)
+		if err != nil {
+			return nil, err
+		}
+
+		return e, nil
+	}
+
 	results := make([]CaseResult, len(t.Cases))
 
 	for i, c := range t.Cases {
@@ -187,7 +205,7 @@ func (r *Runner) runCases(ctx context.Context, suiteDir string, filter Filter, t
 			continue
 		}
 
-		results[i] = r.runCase(ctx, newClient, suiteDir, c)
+		results[i] = r.runCase(ctx, newClient, newExpander, suiteDir, c)
 	}
 
 	return results, nil
@@ -214,6 +232,18 @@ func (r *Runner) makeTestClient(ctx context.Context, suiteDir string, t *Test) (
 	}
 
 	return client, nil
+}
+
+func (r *Runner) makeTestExpander(suiteDir string, t *Test) (*expand.Expander, error) {
+	// Support Mutator logic? Then we need to add support for mutators as well or do we just ignore them?
+	expansionPath := t.Expansion
+	if expansionPath == "" {
+		return nil, nil
+	}
+
+	et, err := reader.ReadExpansion(r.filesystem, path.Join(suiteDir, expansionPath))
+	er, err := expand.NewExpander([]*unstructured.Unstructured{et})
+	return er, err
 }
 
 func (r *Runner) addConstraint(ctx context.Context, suiteDir, constraintPath string, client gator.Client) error {
@@ -252,9 +282,9 @@ func (r *Runner) addTemplate(suiteDir, templatePath string, client gator.Client)
 }
 
 // RunCase executes a Case and returns the result of the run.
-func (r *Runner) runCase(ctx context.Context, newClient func() (gator.Client, error), suiteDir string, tc *Case) CaseResult {
+func (r *Runner) runCase(ctx context.Context, newClient func() (gator.Client, error), newExpander func() (*expand.Expander, error), suiteDir string, tc *Case) CaseResult {
 	start := time.Now()
-	trace, err := r.checkCase(ctx, newClient, suiteDir, tc)
+	trace, err := r.checkCase(ctx, newClient, newExpander, suiteDir, tc)
 
 	return CaseResult{
 		Name:    tc.Name,
@@ -264,7 +294,7 @@ func (r *Runner) runCase(ctx context.Context, newClient func() (gator.Client, er
 	}
 }
 
-func (r *Runner) checkCase(ctx context.Context, newClient func() (gator.Client, error), suiteDir string, tc *Case) (trace *string, err error) {
+func (r *Runner) checkCase(ctx context.Context, newClient func() (gator.Client, error), newExpander func() (*expand.Expander, error), suiteDir string, tc *Case) (trace *string, err error) {
 	if tc.Object == "" {
 		return nil, fmt.Errorf("%w: must define object", gator.ErrInvalidCase)
 	}
@@ -274,7 +304,7 @@ func (r *Runner) checkCase(ctx context.Context, newClient func() (gator.Client, 
 		return nil, fmt.Errorf("%w: assertions must be non-empty", gator.ErrInvalidCase)
 	}
 
-	review, err := r.runReview(ctx, newClient, suiteDir, tc)
+	review, err := r.runReview(ctx, newClient, newExpander, suiteDir, tc)
 	if err != nil {
 		return nil, err
 	}
@@ -293,8 +323,13 @@ func (r *Runner) checkCase(ctx context.Context, newClient func() (gator.Client, 
 	return trace, nil
 }
 
-func (r *Runner) runReview(ctx context.Context, newClient func() (gator.Client, error), suiteDir string, tc *Case) (*types.Responses, error) {
+func (r *Runner) runReview(ctx context.Context, newClient func() (gator.Client, error), newExpander func() (*expand.Expander, error), suiteDir string, tc *Case) (*types.Responses, error) {
 	c, err := newClient()
+	if err != nil {
+		return nil, err
+	}
+
+	e, err := newExpander()
 	if err != nil {
 		return nil, err
 	}
@@ -327,7 +362,32 @@ func (r *Runner) runReview(ctx context.Context, newClient func() (gator.Client, 
 		Object: *toReview,
 		Source: mutationtypes.SourceTypeOriginal,
 	}
-	return c.Review(ctx, au, reviews.EnforcementPoint(util.GatorEnforcementPoint))
+
+	review, err := c.Review(ctx, au, reviews.EnforcementPoint(util.GatorEnforcementPoint))
+
+	if e != nil {
+		resultants, err := e.Expand(toReview)
+		if err != nil {
+			return nil, fmt.Errorf("expanding resource %s: %w", toReview.GetName(), err)
+		}
+
+		for _, resultant := range resultants {
+			au := target.AugmentedUnstructured{
+				Object: *resultant.Obj,
+				Source: mutationtypes.SourceTypeGenerated,
+			}
+			resultantReview, err := c.Review(ctx, au, reviews.EnforcementPoint(util.GatorEnforcementPoint))
+			if err != nil {
+				return nil, fmt.Errorf("reviewing expanded resource %v %s/%s: %w",
+					resultant.Obj.GroupVersionKind(), resultant.Obj.GetNamespace(), resultant.Obj.GetName(), err)
+			}
+			expansion.OverrideEnforcementAction(resultant.EnforcementAction, resultantReview)
+			expansion.AggregateResponses(resultant.TemplateName, review, resultantReview)
+			expansion.AggregateStats(resultant.TemplateName, review, resultantReview)
+		}
+	}
+
+	return review, err
 }
 
 func (r *Runner) validateAndReviewAdmissionReviewRequest(ctx context.Context, c gator.Client, toReview *unstructured.Unstructured) (*types.Responses, error) {

--- a/pkg/gator/verify/runner.go
+++ b/pkg/gator/verify/runner.go
@@ -144,11 +144,6 @@ func (r *Runner) runTest(ctx context.Context, suiteDir string, filter Filter, t 
 	}
 }
 
-func (r *Runner) tryAddExpansion(suiteDir string, t *Test) error {
-	_, err := r.makeTestExpander(suiteDir, t)
-	return err
-}
-
 func (r *Runner) tryAddConstraint(ctx context.Context, suiteDir string, t *Test) error {
 	client, err := r.newClient(r.includeTrace, r.useK8sCEL)
 	if err != nil {
@@ -245,6 +240,10 @@ func (r *Runner) makeTestExpander(suiteDir string, t *Test) (*expand.Expander, e
 	}
 
 	et, err := reader.ReadExpansion(r.filesystem, path.Join(suiteDir, expansionPath))
+	if err != nil {
+		return nil, err
+	}
+
 	er, err := expand.NewExpander([]*unstructured.Unstructured{et})
 	return er, err
 }

--- a/pkg/gator/verify/runner.go
+++ b/pkg/gator/verify/runner.go
@@ -366,6 +366,10 @@ func (r *Runner) runReview(ctx context.Context, newClient func() (gator.Client, 
 	}
 
 	review, err := c.Review(ctx, au, reviews.EnforcementPoint(util.GatorEnforcementPoint))
+	if err != nil {
+		return nil, fmt.Errorf("reviewing %v %s/%s: %w",
+			toReview.GroupVersionKind(), toReview.GetNamespace(), toReview.GetName(), err)
+	}
 
 	if e != nil {
 		resultants, err := e.Expand(toReview)

--- a/pkg/gator/verify/runner_test.go
+++ b/pkg/gator/verify/runner_test.go
@@ -1170,6 +1170,68 @@ func TestRunner_Run(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "expansion system",
+			suite: Suite{
+				Tests: []Test{
+					{
+						Name:       "check custom field with expansion system",
+						Template:   "template.yaml",
+						Constraint: "constraint.yaml",
+						Expansion:  "expansion.yaml",
+						Cases: []*Case{
+							{
+								Name:       "Foo Template object",
+								Object:     "foo-template.yaml",
+								Assertions: []Assertion{{Message: ptr.To[string]("foo object has restricted custom field")}},
+							},
+						},
+					},
+					{
+						Name:       "check custom field without expansion system",
+						Template:   "template.yaml",
+						Constraint: "constraint.yaml",
+						Cases: []*Case{
+							{
+								Name:       "Foo Template object",
+								Object:     "foo-template.yaml",
+								Assertions: []Assertion{{Violations: gator.IntStrFromStr("no")}},
+							},
+						},
+					},
+				},
+			},
+			f: fstest.MapFS{
+				"template.yaml": &fstest.MapFile{
+					Data: []byte(fixtures.TemplateRestrictCustomField),
+				},
+				"constraint.yaml": &fstest.MapFile{
+					Data: []byte(fixtures.ConstraintRestrictCustomField),
+				},
+				"foo-template.yaml": &fstest.MapFile{
+					Data: []byte(fixtures.ObjectFooTemplate),
+				},
+				"expansion.yaml": &fstest.MapFile{
+					Data: []byte(fixtures.ExpansionRestrictCustomField),
+				},
+			},
+			want: SuiteResult{
+				TestResults: []TestResult{
+					{
+						Name: "check custom field with expansion system",
+						CaseResults: []CaseResult{
+							{Name: "Foo Template object"},
+						},
+					},
+					{
+						Name: "check custom field without expansion system",
+						CaseResults: []CaseResult{
+							{Name: "Foo Template object"},
+						},
+					},
+				},
+			},
+		},
 	}
 
 	for _, tc := range testCases {

--- a/pkg/gator/verify/suite.go
+++ b/pkg/gator/verify/suite.go
@@ -35,6 +35,10 @@ type Test struct {
 	// the Suite. Must be an instance of Template.
 	Constraint string `json:"constraint"`
 
+	// Expansion is the path to the Expansion, relative to the file defining
+	// the Suite.
+	Expansion string `json:"expansion"`
+
 	// Cases are the test cases to run on the instantiated Constraint.
 	// Mutually exclusive with Invalid.
 	Cases []*Case `json:"cases,omitempty"`

--- a/test/gator/verify/allow_expansion.yaml
+++ b/test/gator/verify/allow_expansion.yaml
@@ -1,0 +1,8 @@
+apiVersion: apps/v1
+kind: FooTemplate
+metadata:
+  name: foo-template
+spec:
+  template:
+    foo: bar
+

--- a/test/gator/verify/deny_expansion.yaml
+++ b/test/gator/verify/deny_expansion.yaml
@@ -1,0 +1,8 @@
+apiVersion: apps/v1
+kind: FooTemplate
+metadata:
+  name: foo-template
+spec:
+  template:
+    foo: qux
+

--- a/test/gator/verify/expansion.yaml
+++ b/test/gator/verify/expansion.yaml
@@ -1,0 +1,15 @@
+apiVersion: expansion.gatekeeper.sh/v1alpha1
+kind: ExpansionTemplate
+metadata:
+  name: expand-foo
+spec:
+  applyTo:
+  - groups: [ "apps" ]
+    kinds: [ "FooTemplate" ]
+    versions: [ "v1" ]
+  templateSource: "spec.template"
+  generatedGVK:
+    kind: "FooIsBar"
+    group: ""
+    version: "v1"
+

--- a/test/gator/verify/suite.yaml
+++ b/test/gator/verify/suite.yaml
@@ -37,3 +37,17 @@ tests:
     object: deny.yaml
     assertions:
     - violations: no
+- name: foo-is-bar-expansion
+  template: template.yaml
+  constraint: constraint.yaml
+  expansion: expansion.yaml
+  cases:
+  - name: foo-bar
+    object: allow_expansion.yaml
+    assertions:
+    - violations: no
+  - name: foo-not-bar
+    object: deny_expansion.yaml
+    assertions:
+    - violations: yes
+

--- a/website/docs/gator.md
+++ b/website/docs/gator.md
@@ -128,7 +128,7 @@ gator test --filename=manifests-and-policies/ --output=json
 `gator verify` organizes tests into three levels: Suites, Tests, and Cases:
 
 - A Suite is a file which defines Tests.
-- A Test declares a ConstraintTemplate, a Constraint, and Cases to test the
+- A Test declares a ConstraintTemplate, a Constraint, an ExpansionTemplate (optional), and Cases to test the
   Constraint.
 - A Case defines an object to validate and whether the object is expected to
   pass validation.
@@ -161,6 +161,8 @@ A Test compiles a ConstraintTemplate, and instantiates a Constraint for the
 ConstraintTemplate. It is an error for the Constraint to have a different type
 than that defined in the ConstraintTemplate spec.crd.spec.names.kind, or for the
 ConstraintTemplate to not compile.
+
+A Test can also optionally compile an ExpansionTemplate.
 
 ### Cases
 
@@ -262,6 +264,25 @@ the `run` flag:
 
 ```shell
 gator verify path/to/suites/... --run "disallowed"
+```
+
+### Validating Generated Resources with ExpansionTemplates
+`gator verify` may be used along with expansion templates to validate generated resources. The expansion template is optionally declared at the test level. If an expansion template is set for a test, gator will attempt to expand each object under the test. The violations for the parent object & its expanded resources will be aggregated.
+
+Example for declaring an expansion template in a Gator Suite:
+```yaml
+apiVersion: test.gatekeeper.sh/v1alpha1
+kind: Suite
+tests:
+- name: expansion
+  template: template.yaml
+  constraint: constraint.yaml
+  expansion: expansion.yaml
+  cases:
+  - name: example-expand
+    object: deployment.yaml
+    assertions:
+    - violations: yes
 ```
 
 ### Validating Metadata-Based Constraint Templates

